// react-bootstrap-typeahead/es/components/Hint.react.js
import _extends from "@babel/runtime/helpers/extends";
import invariant from 'invariant';
import React, { cloneElement, useCallback, useEffect, useRef } from 'react';
import { useTypeaheadContext } from '../core/Context';
import { shouldSelectHint } from '../utils'; // IE doesn't seem to get the composite computed value (eg: 'padding',
// 'borderStyle', etc.), so generate these from the individual values.

function interpolateStyle(styles, attr, subattr) {
  if (subattr === void 0) {
    subattr = '';
  }

  // Title-case the sub-attribute.
  if (subattr) {
    /* eslint-disable-next-line no-param-reassign */
    subattr = subattr.replace(subattr[0], subattr[0].toUpperCase());
  }

  return ['Top', 'Right', 'Bottom', 'Left'].map(function (dir) {
    return styles[attr + dir + subattr];
  }).join(' ');
}

function copyStyles(inputNode, hintNode) {
  if (!inputNode || !hintNode) {
    return;
  }

  var inputStyle = window.getComputedStyle(inputNode);
  /* eslint-disable no-param-reassign */

  hintNode.style.borderStyle = interpolateStyle(inputStyle, 'border', 'style');
  hintNode.style.borderWidth = interpolateStyle(inputStyle, 'border', 'width');
  hintNode.style.fontSize = inputStyle.fontSize;
  hintNode.style.height = inputStyle.height;
  hintNode.style.lineHeight = inputStyle.lineHeight;
  hintNode.style.margin = interpolateStyle(inputStyle, 'margin');
  hintNode.style.padding = interpolateStyle(inputStyle, 'padding');
  /* eslint-enable no-param-reassign */
}

var Hint = function Hint(_ref) {
  var children = _ref.children,
      className = _ref.className;
  var context = useTypeaheadContext();
  var hintText = context.hintText,
      initialItem = context.initialItem,
      inputNode = context.inputNode,
      onAdd = context.onAdd;
  var hintRef = useRef(null);
  !(React.Children.count(children) === 1) ? process.env.NODE_ENV !== "production" ? invariant(false, 'The `Hint` component expects one child.') : invariant(false) : void 0;
  var onKeyDown = useCallback(function (e) {
    if (shouldSelectHint(e, context)) {
      e.preventDefault(); // Prevent input from blurring on TAB.

      initialItem && onAdd(initialItem);
    }

    children.props.onKeyDown(e);
  });
  useEffect(function () {
    copyStyles(inputNode, hintRef.current);
  });
  return React.createElement("div", {
    className: className,
    style: {
      display: 'flex',
      flex: 1,
      height: '100%',
      position: 'relative'
    }
  }, cloneElement(children, _extends({}, children.props, {
    onKeyDown: onKeyDown
  })), React.createElement("input", {
    "aria-hidden": true,
    className: "rbt-input-hint",
    ref: hintRef,
    readOnly: true,
    style: {
      backgroundColor: 'transparent',
      borderColor: 'transparent',
      boxShadow: 'none',
      color: 'rgba(0, 0, 0, 0.35)',
      left: 0,
      pointerEvents: 'none',
      position: 'absolute',
      top: 0,
      width: '100%'
    },
    tabIndex: -1,
    value: hintText
  }));
};

export default Hint;
