// react-bootstrap-typeahead/es/core/TypeaheadManager.js
import _extends from "@babel/runtime/helpers/extends";
import React, { useCallback, useEffect, useRef } from 'react';
import { TypeaheadContext } from './Context';
import { getHintText, getInputProps, getInputText, getIsOnlyResult, pick } from '../utils';
import { RETURN } from '../constants';
var inputPropKeys = ['activeIndex', 'disabled', 'id', 'inputRef', 'isFocused', 'isMenuShown', 'multiple', 'onBlur', 'onChange', 'onFocus', 'onKeyDown', 'placeholder'];
var propKeys = ['activeIndex', 'isMenuShown', 'labelKey', 'onClear', 'onHide', 'onRemove', 'results', 'selected', 'text'];
var contextKeys = ['activeIndex', 'id', 'initialItem', 'inputNode', 'items', 'onActiveItemChange', 'onAdd', 'onInitialItemChange', 'onMenuItemClick', 'selectHintOnEnter'];

function usePrevious(value) {
  var ref = useRef(null);
  useEffect(function () {
    ref.current = value;
  });
  return ref.current;
}

var TypeaheadManager = function TypeaheadManager(props) {
  var allowNew = props.allowNew,
      children = props.children,
      initialItem = props.initialItem,
      isMenuShown = props.isMenuShown,
      onAdd = props.onAdd,
      onInitialItemChange = props.onInitialItemChange,
      onKeyDown = props.onKeyDown,
      onMenuToggle = props.onMenuToggle,
      results = props.results;
  var prevProps = usePrevious(props);
  useEffect(function () {
    // Clear the initial item when there are no results.
    if (!(allowNew || results.length)) {
      onInitialItemChange(null);
    }
  });
  useEffect(function () {
    if (prevProps && prevProps.isMenuShown !== isMenuShown) {
      onMenuToggle(isMenuShown);
    }
  });
  var handleKeyDown = useCallback(function (e) {
    switch (e.keyCode) {
      case RETURN:
        if (initialItem && getIsOnlyResult(props)) {
          onAdd(initialItem);
        }

        break;

      default:
        break;
    }

    onKeyDown(e);
  });
  var value = getInputText(props);

  var childProps = _extends({}, pick(props, propKeys), {
    getInputProps: getInputProps(_extends({}, pick(props, inputPropKeys), {
      onKeyDown: handleKeyDown,
      value: value
    }))
  });

  var contextValue = _extends({}, pick(props, contextKeys), {
    hintText: getHintText(props),
    isOnlyResult: getIsOnlyResult(props),
    value: value
  });

  return React.createElement(TypeaheadContext.Provider, {
    value: contextValue
  }, children(childProps));
};

export default TypeaheadManager;
